### DISCUSSION TITLE
Using VanillaJS Datepicker with Livewire/AlpineJS

### DISCUSSION THREAD

**Discussion by jrousan** [livewire/livewire#5465] [1. Help] [answered]
I'm attempting to use VanillaJS in my project (https://mymth.github.io/vanillajs-datepicker/#/)

In a blade component, I've created the input component, which stripped down is something like:

```
    <div
        x-data="{ value: @entangle($attributes->wire('model'))}"
        x-init="
        new Datepicker($refs.input,
            {
                autohide: true,
                startView: 2,
                maxDate: new Date(),
             })"
        x-on:change="value = $event.target.value"
    >    
            <div>
                <input
                    {{ $attributes->whereDoesntStartWith('wire:model') }}
                    {{ $attributes->class([...])}}
                    x-ref="input"
                    x-bind:value="value"
                    type="text" autocomplete="off" />
            </div>
           <div>
            @error($name)
            <p {{ $attributes->merge(['class' => 'mt-2 text-sm text-negative-600']) }}>
                {{ $message }}
            </p>
            @enderror
        </div>        
    </div>
```

This works to enable the datepicker on the field.  The problem seems to be that the JavaScript library adds the <div> element for the datepicker directly under the input field.  Thus on any Livewire requests, the datepicker disappears completely.  Adding wire:ignore to the input works, but then I do not get validation class changes.

Is there some sort of workaround for this?  Is there a way to add wire:ignore to dynamically placed DOM elements or am I just SOL with this library?

**Comment by jrousan** [comment, marked as answer]
I've done a workaround like this:

```
    <script>
        const observer = new MutationObserver(function(mutations) {
            mutations.forEach(function(mutation) {
                if (mutation.addedNodes.length) {
                    const node = mutation.addedNodes[0]
                    if (node.classList.contains('datepicker')) {
                        document.getElementById('datepicker').appendChild(node)
                    }
                }
            })
        })

        observer.observe(document.getElementById('datepicker').parentElement, {
            childList: true,
        })
    </script>
```

I created a div with an id of 'datepicker' w/ wire:ignore, then used an observer to look for the datepicker div and append it as a child.  For some reason, adding an attribute of wire:ignore on the datepicker div itself didn't work.

Would still like insight, if there's a better way to handle something like this.

**Comment by joshhanley** [comment]
@jrousan you could try adding `wire:ignore` to the `<div>` that is surrounding the input, instead of on the datepicker div itself, that way everything else should still function.

Like
```blade
<div
        x-data="{ value: @entangle($attributes->wire('model'))}"
        x-init="
        new Datepicker($refs.input,
            {
                autohide: true,
                startView: 2,
                maxDate: new Date(),
             })"
        x-on:change="value = $event.target.value"
    >    
            <div wire:ignore> <!-- wire:ignore here so everything outside of it still works as expected -->
                <input
                    {{ $attributes->whereDoesntStartWith('wire:model') }}
                    {{ $attributes->class([...])}}
                    x-ref="input"
                    x-bind:value="value"
                    type="text" autocomplete="off" />
            </div>
           <div>
            @error($name)
            <p {{ $attributes->merge(['class' => 'mt-2 text-sm text-negative-600']) }}>
                {{ $message }}
            </p>
            @enderror
        </div>        
    </div>
```

Hope this helps!

**Reply by jrousan** [reply]
This does work, but there are conditional classes being applied when there is an error which doesn't work when wire:ignore is applied so I still haven't found a perfect work around that doesn't include an observer.

**Reply by joshhanley** [reply]
@jrousan when you `wire:ignore` any of the conditional classes won't update if rendered by blade. Instead you will need to change the conditional classes to be controlled by Alpine, which should then work happily.

I've built a bunch of components like this and never needed to use an observer, so I'm sure we can come up with a way around it 🙂
